drakrun/drakrun/draksetup.py

```diff
@@ -75,6 +76,79 @@ def ensure_zfs(ctx, param, value):
     return value
 
 
+def sanity_check():
+    logging.info("Checking xen-detect...")
+    proc = subprocess.run('xen-detect -N', shell=True)
+
+    if proc.returncode != 1:
+        logging.error('It looks like the system is not running on Xen. Please reboot your machine into Xen hypervisor.')
+        return
+
+    logging.info("Testing if xl tool is sane...")
+
+    try:
+        subprocess.check_output('xl info', shell=True, stderr=subprocess.STDOUT)
+    except subprocess.CalledProcessError:
+        logging.exception("Failed to test xl command.")
+        return
+
+    if not perform_xtf():
+        logging.error("Your Xen installation doesn\'t pass the necessary tests.")
+        return
+
+
+def perform_xtf():
+    logging.info('Testing your Xen installation...')
+    module_dir = os.path.dirname(os.path.realpath(__file__))
+    cfg_path = os.path.join(module_dir, "tools", "test-hvm64-example.cfg")
+    firmware_path = os.path.join(module_dir, "tools", "test-hvm64-example")
+
+    with open(cfg_path, 'r') as f:
+        test_cfg = f.read().replace('{{ FIRMWARE_PATH }}', firmware_path).encode('utf-8')
+
+    with tempfile.NamedTemporaryFile() as tmpf:
+        tmpf.write(test_cfg)
+        tmpf.flush()
+
+        try:
+            subprocess.check_output('xl destroy test-hvm64-example', shell=True, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError:
+            pass
+
+        subprocess.check_output(f'xl create -p {tmpf.name}', shell=True, stderr=subprocess.STDOUT, timeout=30)
+
+        module_dir = os.path.dirname(os.path.realpath(__file__))
+        test_altp2m_tool = os.path.join(module_dir, "tools", "test-altp2m")
+
+        try:
+            output = subprocess.check_output([test_altp2m_tool, 'test-hvm64-example'], stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            output = e.output.decode('utf-8', 'replace')
+            logging.error(f'Failed to enable altp2m on domain. Your hardware might not support Extended Page Tables. Logs:\n{output}')
+            return False
+
+        p = subprocess.Popen(['xl', 'console', 'test-hvm64-example'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        subprocess.check_output('xl unpause test-hvm64-example', shell=True, stderr=subprocess.STDOUT, timeout=30)
+        stdout_b, _ = p.communicate(timeout=10)
+
+        stdout_text = stdout_b.decode('utf-8')
+        stdout = [line.strip() for line in stdout_text.split('\n')]
+
+        for line in stdout:
+            if line == 'Test result: SUCCESS':
```

Comment:
Am I blind or `test-hvm64-example` doesn't print out this exact line?

Reply:
Have you reversed the firmware binary or what? 🤔

Reply:
This binary was built out of XTF:
https://xenbits.xen.org/gitweb/?p=xtf.git;a=blob;f=tests/example/main.c;h=32422f1439bf498fb0c7468dc7e1906da9c6d85e;hb=HEAD

at the moment we just include the binary instead of having the proper build process in our scripts. These tests are implemented as a part of some test framework. This is the simplest possible test, so it just prints `Hello World` and returns "success".

Reply:
Ummmm, no, there was no need to
![image](https://user-images.githubusercontent.com/8405359/109502534-68193700-7a99-11eb-877d-36d0bc8a8037.png)

Reply:
This is a bug in GitHub, this file is not empty o_O

Reply:
Apparently file is there, but github is acting weird.

> Have you reversed the firmware binary or what?

ida goes brrr